### PR TITLE
Add Translations Dashboard Panel (v1)

### PR DESCRIPTION
**Pull request by timarney**
# Summary | Résumé

This is a still-needs-polishing but functional implementation of a translations dashboard widget that allows for changing the assigned translation for a given post or page.

Screen Shot 2022-07-28 at 14 58 20
 
Things that work:
- [3 API routes](https://github.com/cds-snc/gc-articles/pull/980#issuecomment-1192901535): 
  1. GET all pages or posts in either English or French
  2. GET one post/page and its associated translation
  3. POST assign a translation to a post/page
- Dashboard widget is functional: can assign a translation and then use "Update" button to save
  - After updating, the widget refreshes its state
- Display posts and pages as either "translated" or "untranslated"
- Hint text more or less knows what's going on
- Lots of tests for the API
- Return up to 100 posts or pages
- Any users with the `edit_post` capability can use this

Things that don't work yet:
- Assigning a "None" translation (essentially, removing a translation from a page that currently has one)
- ~Not all pages are not returned by the API but I am not sure why yet~
- ~No nonce being used for the POST request, so anyone can reassign posts if they know how to (if they pass validation)~
- Adding/checking French translations

Basically, what we have here is something that Phil and Charlotte can work with while we (I) continue to make it a bit more robust.

### gif

Screen Recording 2022-07-28 at 14 54 22

### CODE REVIEW

**Comment by pcraig3** [comment]
An update from me, I've filled in the 3 API routes given.

1. GET all posts / pages for a given language
2. GET a translation for a given post id 
3. POST an update linking a given post ID to another post ID (the one in the path is considered the "source")

### 1. GET all posts / pages for a given language

Use "posts" or "pages" in the path, and "en" or "fr" at the end.

URL: [`http://localhost/wp-json/cds/wpml/posts/fr`](http://localhost/wp-json/cds/wpml/posts/fr)

<details>
<summary>Example response: GET all French posts</summary>

```json
[
   {
      "ID":49,
      "post_title":"Post: 49 (FR)",
      "post_type":"post",
      "language_code":"fr",
      "translated_post_id":45,
      "is_translated":true
   },
   {
      "ID":47,
      "post_title":"Post: 47 (FR)",
      "post_type":"post",
      "language_code":"fr",
      "translated_post_id":53,
      "is_translated":true
   },
   {
      "ID":32,
      "post_title":"Post: Charles-Maurice de Talleyrand-Périgord (FR)",
      "post_type":"post",
      "language_code":"fr",
      "translated_post_id":null,
      "is_translated":false
   }
]

```

</details>


### 2. GET a translation for a given post id 

Only "posts" is allowed here, and use the id of the post in the path. Returns info on the translated post, or a 404 error message.

URL: [`http://localhost/wp-json/cds/wpml/posts/43/translation`](http://localhost/wp-json/cds/wpml/posts/43/translation)

<details>
<summary>Example response: GET translation for EN post (with translation)</summary>

```json
{
"ID": 49,
"post_title": "Post: 49 (FR)",
"post_type": "post",
"language_code": "fr"
}

```

</details>

<details>

<summary>Example response: GET translation for EN post ( *no* translation exists)</summary>

```json
{
   "code":"no_translation",
   "message":"No translation exists for this post.",
   "data":{
      "status":404
   }
}

```

</details>


### 3. POST an update linking a given post ID to another post ID 

Same URL as the (2), but this is a POST request, and also requires a body param to be sent: `other_id`. I couldn't think of a good name for it, so we should probably change it. 

This one is pretty complex. 

This route is deliberately permissive. There are 4 cases here:

1. New post (no trns) -> other post (no trns)
2. New post (w/ trns) -> other post (no trns)
3. New post (no trns) -> other post (w/ trns)
4. New post (w/ trns) -> other post (w/ trns)

If we prevented scenarios 2, 3, and 4, we would be in a position where you would have no way to reassign a translation once you assigned one, other than deleting your post and starting over.

In the scenarios where a translation is already assigned, the logic will create orphans of the previously-linked posts and make sure the current one is set.

Eg,

Say we have post 1 (EN), post 2 (FR), and post 3 (EN). If post 1 and 2 are linked, but we submit a request to link 3 -> 2, then
post 1 will become an orphan and 3 -> 2 will be linked.


URL: [`http://localhost/wp-json/cds/wpml/posts/47/translation`](http://localhost/wp-json/cds/wpml/posts/47/translation)
- `other_id`: 53 

<details>
<summary>Example response: Add translation (post 53) to post 47</summary>

```json
{
  "ID": 47,
  "post_title": "Post: 47 (FR)",
  "post_type": "post",
  "language_code": "fr",
  "translated_post_id": 53,
  "is_translated": true
}

```

</details>

File: wordpress/wp-content/plugins/cds-wpml-mods/resources/js/util/fetch.js
Comment (timarney) [inline wordpress/wp-content/plugins/cds-wpml-mods/resources/js/util/fetch.js:9]: Used on the auth end of things

See: https://developer.wordpress.org/rest-api/using-the-rest-api/authentication/#cookie-authentication

File: wordpress/wp-content/plugins/cds-wpml-mods/src/Api/Endpoints.php
Comment (timarney) [inline wordpress/wp-content/plugins/cds-wpml-mods/src/Api/Endpoints.php:34]: This needs to be updated before we merge

Comment (timarney) [reply wordpress/wp-content/plugins/cds-wpml-mods/src/Api/Endpoints.php:34]: Can check if the nonce check happens here too

Comment (pcraig3) [reply wordpress/wp-content/plugins/cds-wpml-mods/src/Api/Endpoints.php:34]: 👍

File: wordpress/wp-content/plugins/cds-wpml-mods/resources/js/components/PageSelect.js
Comment (timarney) [inline wordpress/wp-content/plugins/cds-wpml-mods/resources/js/components/PageSelect.js:73]: Thinking it would be better to pass the post attributes we need for this function 

I think that will help prevent refetch / loading again when you  move between the page / block tabs


Screen Shot 2022-07-29 at 9 58 42 AM

Right now if you swap between them there's a fetch happening but there doesn't need to be.

Comment (pcraig3) [reply wordpress/wp-content/plugins/cds-wpml-mods/resources/js/components/PageSelect.js:73]: Ah, I wasn't sure why that was happening, but you're right, there's no reason it needs to.

File: wordpress/wp-content/plugins/cds-wpml-mods/resources/js/components/PageSelect.js
Comment (timarney) [inline wordpress/wp-content/plugins/cds-wpml-mods/resources/js/components/PageSelect.js:42]: Looks like there's still an issue here when setting back to `None` ya?
Screen Shot 2022-07-29 at 10 09 14 AM

**Comment by timarney** [comment]
Overall this is looking really good - some solid improvements to the UI etc... since the last time I checked.

File: wordpress/wp-content/plugins/cds-wpml-mods/resources/js/components/PageSelect.js
Comment (pcraig3) [reply wordpress/wp-content/plugins/cds-wpml-mods/resources/js/components/PageSelect.js:42]: Yeah, there's no API route for it yet. We can ship it without this working assuming nobody needs to do this for a day or two.

**Review by pcraig3 — APPROVED** [review]

**Comment by timarney** [comment]
Approved ✅

**Pull request merged**
